tests/test_circuit.py

```diff
@@ -1,146 +1,171 @@
+        cls.results_circuits = [[], []]
+        for i,c in enumerate(cls.circuits):
+            for j,circuit in enumerate(c):
+                    circuit_function = circuit.build_circuit_function
+                    cls.results_circuits[i].append(
+                        circuit.run_and_measure_circuit(
+                        circuit_function)(cls.input, cls.params[j])
+                    )
 
-
-    def test_circuits_output_correct_type(self):
+    def test_circuits_output_correct_type(self)-> None:
         """
-        Test the that the the three different ansatze implemented 
-        work. To do so it will be asserted that the expected value obtained
-        after running the circuit is a tensor
+        Test the that the three different ansatze implemented 
+        work (for expectation value and prob outputs),
+        asserting that a tensor is obtained as output.
         """
-        for i,circuit in enumerate(self.circuits):
-            circuit_function = circuit.build_circuit_function
-            results = circuit.run_and_measure_circuit(
-                circuit_function)(self.input, self.params[i])
-            ### We make sure the output has the proper format
-            for r in results:
-                self.assertIsInstance(r, torch.Tensor)
+        for i,r in enumerate(self.results_circuits):
+            for result in r:
+                with self.subTest(result=result):
+                    for tensor in result:
+                        self.assertIsInstance(tensor, torch.Tensor)
 
-    def test_result_of_different_rescaling_functions(self):
+    def test_circuit_outputs_probabilities_between_0_and_1(self)-> None:
         """
-        Test the results of rescaling function of the inputs,
-        ensuring that it varies when different 
-        rescaling functions are applied
+        For all the ansatze, test that when probabilities are output,
+        they lay in the range (0,1).
         """
-        data_rescaling_functions = [
-            lambda x : 0.5 * mt.pi * x,
-            lambda x : mt.pi * x,
-            lambda x : 2.0 * mt.pi * x
-        ]
-        for i,ansatz in enumerate(self.ansatze):
-            results_list = []
-            for f in data_rescaling_functions:
-                circuit = ansatz(
-                    self.n_qubits, self.n_layers,
-                    self.axis_encoding, self.observables,
-                    data_rescaling = f
-                )
-                circuit_function = circuit.build_circuit_function
-                results = circuit.run_and_measure_circuit(
-                circuit_function)(self.input, self.params[i])
-                results_list.append(results)
-            self.assertNotEqual(results_list[0], results_list[1])
-            self.assertNotEqual(results_list[1], results_list[2])
-            self.assertNotEqual(results_list[2], results_list[0])
-                          
+        for result in self.results_circuits[1]:
```

Comment:
Instead of all those for loops, wasn't there a way to use `all`? I'm not saying there is necessarily one, just checking that you'v looked into it. I'd assume `all` to work with these objects but never tried it so who knows!

Reply:
Good suggestion ryk, I hadn't thought about that. I've used all function (the one from torch, as we are working with tensors) to eliminate the inner loop. 

However, I wasn't able to figure out how to eliminate the two outer loops with all() function. At most, I could use all() among with list comprehension, something like : 

all(all(x > 0 for x in sublist) for sublist in nested_list). 

But we will still have the loops. I personally think it's clearer as it is rn, but if you want I can change it, having smthing like the line above.

Reply:
I like the all + list comprehension better than the for loop. more pythonic imho

Reply:
After looking this up, it seems that is not possible to do list comprehension and using subTests at the same time (or at least I wasn't able to). If you want I can explain further next DSU.